src/validators/jwks.rs

```diff
@@ -0,0 +1,489 @@
+use crate::{
+	apis::jwks_api::{Jwks, JwksKey},
+	clerk::Clerk,
+	validators::authorizer::ClerkError,
+};
+use arc_swap::{ArcSwap, Guard};
+use async_trait::async_trait;
+use std::{
+	collections::HashMap,
+	sync::Arc,
+	time::{Duration, SystemTime},
+};
+
+#[async_trait]
+pub trait JwksProvider {
+	type Error: Into<ClerkError>;
+
+	async fn get_key(&self, kid: &str) -> Result<JwksKey, Self::Error>;
+}
+
+/// Error type used by [`MemoryCacheJwksProvider`] and [`SimpleJwksProvider`].
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum JwksProviderError {
+	UnknownKey,
+	JwksApi,
+}
+
+impl From<JwksProviderError> for ClerkError {
+	fn from(e: JwksProviderError) -> Self {
+		match e {
+			JwksProviderError::UnknownKey => ClerkError::Unauthorized("Error: Invalid JWT!".into()),
+			JwksProviderError::JwksApi => ClerkError::InternalServerError(String::from("Error: Could not fetch JWKS!")),
+		}
+	}
+}
+
+/// A [`JwksProvider`] implementation that doesn't do any caching.
+/// The JWKS is fetched from the Clerk API on every request.
+pub struct SimpleJwksProvider {
+	clerk_client: Clerk,
+}
+
+impl SimpleJwksProvider {
+	pub fn new(clerk_client: Clerk) -> Self {
+		Self { clerk_client }
+	}
+}
+
+#[async_trait]
+impl JwksProvider for SimpleJwksProvider {
+	type Error = JwksProviderError;
+
+	async fn get_key(&self, kid: &str) -> Result<JwksKey, JwksProviderError> {
+		let jwks = Jwks::get_jwks(&self.clerk_client).await.map_err(|_| JwksProviderError::JwksApi)?;
+
+		jwks.keys.into_iter().find(|k| k.kid == kid).ok_or(JwksProviderError::UnknownKey)
+	}
+}
+
+pub enum RefreshOnUnknown {
+	/// Never attempt to refresh the JWKS when an unknown `kid` is requested.
+	Never,
+	/// Attempt to refresh the JWKS when an unknown `kid` is requested
+	/// if the cached value is older than the given duration.
+	Ratelimit(Duration),
+	/// Always attempt to refresh the JWKS when an unknown `kid` is requested.
+	Always,
+}
+
+pub struct MemoryCacheJwksProviderOptions {
```

Comment:
Is it possible to add some doc comments for the struct itself?

Reply:
Added some new short comments to the structs, I think the comments on the fields are pretty explanatory though